src/ucp/rma/rma_send.c

```diff
@@ -127,7 +127,17 @@ ucp_rma_request_init(ucp_request_t *req, ucp_ep_h ep, const void *buffer,
     req->send.ep              = ep;
     req->send.buffer          = (void*)buffer;
     req->send.datatype        = ucp_dt_make_contig(1);
-    req->send.mem_type        = UCS_MEMORY_TYPE_HOST;
+    req->send.mem_type        = ucp_memory_type_detect(ep->worker->context,
+                                                       buffer, length);
+    if (ucs_unlikely((req->send.mem_type != UCS_MEMORY_TYPE_HOST) ||
+                     (rkey->mem_type != UCS_MEMORY_TYPE_HOST))) {
```

Comment:
remote memory **could** be non host in case of GPU direct.. if we got the rkey for, it means it was registered on remote side so it's accessible

Reply:
> remote memory **could** be non host in case of GPU direct.. if we got the rkey for, it means it was registered on remote side so it's accessible

`rkey->mem_type` can be non-HOST (e.g. `UCS_MEMORY_TYPE_CUDA`) in case of `UCX_TLS=tcp,cuda_copy`
do you mean that we have to check `md_map` to understand whether it was registered on network TL (e.g. IB using GPUDirect) or not?
If we see that this memory (local or remote) was registered using GPUDirect, then we could support UCP RMA

Reply:
for tcp case, it's different, it's sw-rma protocol which indeed works only on host memory

Reply:
> for tcp case, it's different, it's sw-rma protocol which indeed works only on host memory

yes
but if we remove checking remote memory type, then we break a RMA operation for `H->C` case, when GPUDirect isn't supported

Reply:
so rma_basic protocol should not check remote memtype, only rma_sw should check remote mem type

Reply:
> so rma_basic protocol should not check remote memtype, only rma_sw should check remote mem type

@yosefe thank you for the clarification!
this is common code for both basic and sw RMA, will move a check for remote memory type to sw RMA only